[scroll-animations] multiple `timeline-scope` WPT tests are failing

graouts opened the pull request:
#### 2f41a237286ad02b6bd743fc2c9f596a0c00ae43
<pre>
[scroll-animations] multiple `timeline-scope` WPT tests are failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=286264">https://bugs.webkit.org/show_bug.cgi?id=286264</a>
<a href="https://rdar.apple.com/143271212">rdar://143271212</a>

Reviewed by Antti Koivisto and Simon Fraser.

When properties of an animation change, we end up calling `Element::invalidateStyleForAnimation()` on the
associated effect target to ensure that those updated animation properties are represented in the animated
style.

The creation of style-originated animations (CSS Animations and CSS Transitions) occur during style resolution,
typically under `Document::resolveStyle()`. During that process, we ensure that the new style-originated
animations do not call into `Element::invalidateStyleForAnimation()` so as not to produce some inconsistent
state.

As part of scroll-driven animations, in the case of `timeline-scope`, we do new work during style resolution to
set the timeline of a style-originated animation. The previous protections we had in place to avoid calling
`Element::invalidateStyleForAnimation()` did not apply there. This would yield inconsistent node states where
the Style::Validity::AnimationInvalid flag would be set but ancestors would no longer have the
`NodeStyleFlag::DescendantNeedsStyleResolution` flag set, preventing further style resolution on animated
elements.

We now add a more direct check in the `invalidateElement()` static function of `KeyframeEffect.cpp` to check
that we do not call `Element::invalidateStyleForAnimation()` during style resolution. Furthermore, we also
add an `ASSERT()` in `Element::invalidateStyleForAnimation()` to check that this holds true for any future
call to that function.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-multi-pass.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-range-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-range-animation-expected.txt:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::invalidateElement):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invalidateStyleForAnimation):

Canonical link: <a href="https://commits.webkit.org/289180@main">https://commits.webkit.org/289180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d046a2d76c5159379a235befc821a52b4c0223d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39953 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90642 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36557 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87603 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5442 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13229 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66483 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24298 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4179 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77679 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46766 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4047 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31942 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-008.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35627 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74723 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32793 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92281 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12872 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9438 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75183 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13090 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73513 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74319 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18576 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17014 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4971 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13347 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12890 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18273 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12692 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16127 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14470 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-expected.txt:
```diff
@@ -1,12 +1,12 @@
 PASS Inner timeline does not interfere with outer timeline
-PASS Deferred timeline with two attachments
-FAIL Dynamically re-attaching assert_equals: expected "100px" but got "0px"
+FAIL Deferred timeline with two attachments assert_equals: expected "0px" but got "100px"
```
This is a regression

graouts replied:
Yeah, it is. I’m fairly certain this source change is absolutely correct as-is but something else is required to get this test working which likely worked by chance. I’ll take a closer look at this specific subtest.

graouts replied:
Indeed, this newly-failing test only worked by virtue of the very bug this patch is fixing. In that test two elements are setting `scroll-timeline-name` to `--t1`, a scrolling element and the animation target. That animation target is also using `--t1` as its `animation-timeline`, but since it's not scrollable that animation target has no valid timeline. That test checks that, as such, the animation doesn't apply.

In our implementation, currently, we use the first element with `scroll-timeline-name` set to `--t1` as the source, and that element is scrollable and thus produces a valid timeline and a resolved current time.

Prior to this patch, the fact that there was a valid timeline and a resolved current time for the animation target was hidden by the fact that we failed to reflect the effect of an animation with a scoped timeline through the computed style, which is the method used by this test.

So while this test is now a `FAIL`, it only ever worked due to this bug masking this other bug. I will land this patch as-is and investigate the bug preventing this test from working.

graouts replied:
This specific test is being fixed in https://github.com/WebKit/WebKit/pull/39324.

graouts committed:
[scroll-animations] multiple `timeline-scope` WPT tests are failing
https://bugs.webkit.org/show_bug.cgi?id=286264
rdar://143271212

Reviewed by Antti Koivisto and Simon Fraser.

When properties of an animation change, we end up calling `Element::invalidateStyleForAnimation()` on the
associated effect target to ensure that those updated animation properties are represented in the animated
style.

The creation of style-originated animations (CSS Animations and CSS Transitions) occur during style resolution,
typically under `Document::resolveStyle()`. During that process, we ensure that the new style-originated
animations do not call into `Element::invalidateStyleForAnimation()` so as not to produce some inconsistent
state.

As part of scroll-driven animations, in the case of `timeline-scope`, we do new work during style resolution to
set the timeline of a style-originated animation. The previous protections we had in place to avoid calling
`Element::invalidateStyleForAnimation()` did not apply there. This would yield inconsistent node states where
the Style::Validity::AnimationInvalid flag would be set but ancestors would no longer have the
`NodeStyleFlag::DescendantNeedsStyleResolution` flag set, preventing further style resolution on animated
elements.

We now add a more direct check in the `invalidateElement()` static function of `KeyframeEffect.cpp` to check
that we do not call `Element::invalidateStyleForAnimation()` during style resolution. Furthermore, we also
add an `ASSERT()` in `Element::invalidateStyleForAnimation()` to check that this holds true for any future
call to that function.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-multi-pass.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-range-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-range-animation-expected.txt:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::invalidateElement):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invalidateStyleForAnimation):

Canonical link: https://commits.webkit.org/289180@main
